receiver/prometheusexec/README.md

```diff
@@ -0,0 +1,126 @@
+# This receiver is currently under development, do not use.
+// TODO: Remove above message when done
+
+# Prometheus_exec Receiver
+
+This receiver makes it easy for a user to collect metrics with the Collector from third-party services such as MySQL, Apache, Redis, etc. It's meant for people who want a plug-and-play solution to getting metrics from those third-party services that don't natively export metrics or speak any instrumentation protocols. Through the configuration file, you can indicate which binaries to run (usually [Prometheus exporters](https://prometheus.io/docs/instrumenting/exporters/), which are custom binaries that will expose the services' metrics to the Prometheus protocol). It supports starting binaries with flags and environment variables, retrying them with exponentional backoff if they crash, string templating, and random port assignments. If you do not need to spawn the binaries locally, please condider using the [core Prometheus receiver](https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/prometheusreceiver) or the [Simple Prometheus receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/receiver/simpleprometheusreceiver).
+
+## Overview
+Here's an example Collector configuration file that starts two `prometheus_exec` receivers and two exporters.
+
+```yaml
+receivers:
+    prometheus_exec/mysqld:
+        exec: ./mysqld_exporter --port {{port}}
+        scrape_interval: 10s
+        port: 9104
+        env:
+          - name: DATA_SOURCE_NAME
+            value: user:password@(url:port)/
+
+    prometheus_exec/postgresql:
+        exec: ./postgres_exporter
+        env:
+          - name: DATA_SOURCE_NAME
+            value: postgresql://user:password@url:port/
+
+exporters:
+    stackdriver:
+        project: "project_name"
+    file:
+        path: ./file_name.json
+
+service:
+    pipelines:
+        metrics:
+            receivers: [prometheus_exec/mysqld, prometheus_exec/postgresql]
+            exporters: [stackdriver, file]
+```
+
+## Config
+For every `prometheus_exec` defined in the configuration file, a command will be run (you *should* start a binary for this receiver to serve its purpose) and an equivalent Prometheus receiver will be instantiated to scrape its metrics.
+
+- ### prometheus_exec (required)
+`prometheus_exec` receivers should hierarchically be placed under the `receivers` key. You can define as many of these as you want, and they should be named as follows: `prometheus_exec/custom_name`. The `custom_name` should be unique per receiver you define, and is important for logging and error tracing. If no `custom_name` is given (i.e. simply having `prometheus_exec`), one will be generated for you by using the first word in the command to be executed (doing this limits the amount of receivers to 1 though since you can't define multiple receivers with duplicate names, so it's not recommended). Example:
```

Comment:
Isn't the first part (named receivers) just regular Collector behaviour? And why bother with the second part at all? Is this because you always need to create a named Prometheus receiver internally? To keep this simpler, perhaps just make the name mandatory for this receiver then?

Reply:
Not sure what you mean by first part here, but yes, the Collector normally works by having 1 receiver per receiver defined, however I wanted to make it clear since it's different from the "original" Prom receiver since that one works with 1 receiver having multiple `scrape_configs`.

The `custom_name` is mostly to prefix logs from the processes running, to help the user debug or read logs - they'll know which process is spitting out those logs. I just didn't want to make it mandatory in case the user forgets or something. I'm open to making it required, though. I thought just generating one for the user would be nice if they forgot or something!

Reply:
Sorry by first part I meant the first three sentences.

> different from the "original" Prom receiver since that one works with 1 receiver having multiple scrape_configs.

That's not strictly true. Users can create multiple named promethus receivers if they like.

I think you may be re-implementing something that the Collector already supports out of the box (other tha ntaking the name from the command by default). It might not be as well documented as I thought. The [design docs](https://github.com/open-telemetry/opentelemetry-collector/blob/master/docs/design.md#receivers) link to this [Google Doc](https://docs.google.com/document/d/1NeheFG7DmcUYo_h2vLtNRlia9x5wOJMlV4QKEK05FhQ/edit#heading=h.74px12cea3py) which specifies how to named components can be created (search for "type[/name]" in that doc)

Reply:
Oh I see what you mean! In the configuration, I'm just trying to indicate to the user that that's how **this** receiver should be invoked (separate receivers with unique names), which differentiates from how Prometheus receivers are **usually** defined (I understand you can have multiple receivers but the *usual* way seems to be simply multiple `scrape_configs`). To be clear, I'm not re-implementing the `receiver[/custom_name]` structure (I'm using what's already built into the Collector), only stating its necessity here! Also, I'm letting the user know where/how the `custom_name` will be used (logging purposes for now).

I think we're agreeing here: since multiple separate receivers need to be defined to have multiple `prometheus_exec`s running at once and the Collector already enforces unique names for multiple receivers, I'm just letting the user know how they should write their config file, and I'm not re-implementing that!

Does this clear any confusion? Thanks for the input btw, very appreciated. 

- You think I should re-word the this part of the README to include the fact that they need custom names if they have multiple receivers and that that is normal Collector behavior, and then say that this specific receiver uses the `custom_name` for logging?

Reply:
I'm still a little bit confused. If this just uses the default behaviour, then how are you implementing this part? "If no `custom_name` is given (i.e. simply having `prometheus_exec`), one will be generated for you by using the first word in the command to be executed"

Are you planning on overriding the Name by calling `NamedEntity.SetName()` from within the receiver? I'd probably recommend not doing that as that then deviates from how all other components behave.

I think I'd probably remove this entire section, or maybe change it to:

"To run multiple Prometheus exporters, create multiple named `prometheus_exec` receivers: (yaml example)"

Reply:
Oh I see where the confusion is! `custom_name` here is simply used by my receiver as a prefix for logging. Whether the user puts something for `name` in `prometheus_exec[/name]` or not, I won't do anything about that. I won't be changing the internal name the Collector uses for that receiver, I'll simply be generating one for **my** use for this receiver.

To recap, if the user defines a name in `prometheus_exec[/name]`, then I use that as my `custom_name` (used only by my receiver for logging prefixes), and if not I generate one for them.

Do you still think I should remove it?

Reply:
So it sounds like you just want to be able to distinguish between logs from different named receivers? If you think about it, functionality where you include the name, type, etc of the component in the logs would be much more useful if implemented as generic behaviour of the collector. Otherwise, it would need to be implemented for each component separately!

And, it is already implemented as generic behaviour. As long as you use the `logger` passed in to your component via the `factory.CreateMetricsReceiver` (etc) function, that logger has already been configured to log that information. See: https://github.com/open-telemetry/opentelemetry-collector/blob/master/service/builder/pipelines_builder.go#L159

If you check the logs you should see something like this:
```
2020-07-04T13:59:15.553+1000	INFO	receiver/my_receiver.go:213	My log message.	{"component_kind": "receiver", "component_type": "my_receiver_type", "component_name": "my_receiver_name", "datatype": "metrics"}
```

Reply:
I might not be able to use the `logger` for logs coming from the subprocesses, but thanks for the info! I'll look into it more and update the README accordingly. Thanks for the discussion! :)

Reply:
Update: I removed this part of the README and removed the name generation code. As per normal Collector behavior, the user needs unique `custom_names` if they want multiple of the same receiver running at once. This way, a `custom_name` is very likely to be defined by the user. On the off chance that the user only defines one of this receiver **and** does not give it a `custom_name`, the logs won't have a prefix but that's not a problem since only one process will be logging, which makes it easy to know which process it's coming from.

To recap, this part of the README is being removed and the `custom_name` will be defined by the user per normal Collector behavior, no `custom_name` generation here.